Check that the size of a cookie's path is not too large when setting a cookie with the Cookie Store API

RupinMittal opened the pull request:
#### d7472f0d81e69ab8c62df926720430d69fdba732
<pre>
Check that the size of a cookie&apos;s path is not too large when setting a cookie with the Cookie Store API
<a href="https://bugs.webkit.org/show_bug.cgi?id=259571">https://bugs.webkit.org/show_bug.cgi?id=259571</a>

Reviewed by Alex Christensen.

The spec (<a href="https://wicg.github.io/cookie-store/#set-cookie-algorithm)">https://wicg.github.io/cookie-store/#set-cookie-algorithm)</a>
dictates that in the set function, if the byte sequence length of the
path (in UTF8 format) is greater than the maximum attribute value size
(currently 1024 bytes according to
<a href="https://wicg.github.io/cookie-store/#cookie-maximum-attribute-value-size)">https://wicg.github.io/cookie-store/#cookie-maximum-attribute-value-size)</a>,
then the promise should be rejected with a TypeError. This patch adds
this check the same way a previous patch added the same check for a
cookie&apos;s domain. It also adds testing for the length check for both path
and domain.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.js:
(promise_test.async testCase):
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.serviceworker-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):

Canonical link: <a href="https://commits.webkit.org/266410@main">https://commits.webkit.org/266410@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7a7a98909772e229ec524c4a9d17dce8bef1a1ed

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13744 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14058 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14391 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15480 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13053 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13825 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16566 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14139 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15729 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13911 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14529 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11636 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16180 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11818 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12394 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19429 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12893 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12559 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15772 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13089 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10963 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12357 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3346 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16690 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12931 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/Modules/cookie-store/CookieStore.cpp:
```diff
@@ -241,8 +242,14 @@ void CookieStore::set(CookieInit&& options, Ref<DeferredPromise>&& promise)
         if (!cookie.path.endsWith('/'))
             cookie.path = cookie.path + '/';
+
+        if (cookie.path.utf8().length() > maximumAttributeValueSize) {
```
It would be nice if we didn't need to allocate and encode just to get the encoded length.  I started this in https://bugs.webkit.org/show_bug.cgi?id=232962 .  That's out of scope in this PR, but might be worth adding a comment to this and the domain.utf8()

achristensen07 replied:
Also, add a test for this.

achristensen07 replied:
And that test should be added to the WPT repo.

RupinMittal replied:
Ok I've added a comment about the possible encoded length improvement. I've also added testing for the path and domain lengths. Ideally, this would be part of the `cookieStore_set_arguments.https.any.html` test since that's where we test all of the ways of setting a cookie with the different arguments (like path and domain). This is where I've added it for the moment. But I'm not sure if we're allowed to just modify the existing tests like that or how I can add this to WPT?

achristensen07 replied:
Yeah, just modify the test copies in our repo like you did, then make another PR to https://github.com/web-platform-tests/wpt

RupinMittal committed:
Check that the size of a cookie's path is not too large when setting a cookie with the Cookie Store API
https://bugs.webkit.org/show_bug.cgi?id=259571

Reviewed by Alex Christensen.

The spec (https://wicg.github.io/cookie-store/#set-cookie-algorithm)
dictates that in the set function, if the byte sequence length of the
path (in UTF8 format) is greater than the maximum attribute value size
(currently 1024 bytes according to
https://wicg.github.io/cookie-store/#cookie-maximum-attribute-value-size),
then the promise should be rejected with a TypeError. This patch adds
this check the same way a previous patch added the same check for a
cookie's domain. It also adds testing for the length check for both path
and domain.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.js:
(promise_test.async testCase):
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_set_arguments.https.any.serviceworker-expected.txt:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::set):

Canonical link: https://commits.webkit.org/266410@main
